### DISCUSSION TITLE
Support for mixing and outputting surround sound

### DISCUSSION THREAD

**Discussion by Hiradur** [jamulussoftware/jamulus#2994] [Ideas]
I've been wondering if it wouldn't be nice to add surround sound support to the output stage of Jamulus. Often, sessions in Jamulus have enough instruments playing that there would be enough content for a surround mix. As an avid listener, I could imagine myself experimenting with panning instruments in real-time in surround.

I can think of at least three ways to implement this:
1) output every channel individually for further processing by an external application
2) internal pan-pot panning for more than just 2 speakers
3) internal panning in an Ambisonics soundfield

Comments:
1. Would be the most flexible option and possibly easiest to implement. It would leave the mixing to an external application. It would also allow the use of effect plugins specifically designed for surround mixing (e.g. for reverb etc.).

2. Would probably be easy to implement but it is said to have [insufficient localization toward the sides and rear](https://www.ambisonic.net/ambidvd.html).

3. Would make it possible to mix to an arbitrary number of speaker channels. However, the implementation would probably be more complex, since rendering the audio would have to happen in B-Format. For the end-user, it would also be more difficult to set up, requiring a decoding matrix for the speaker layout they use and setting up an Ambisonics decoder (e.g. ambdec) to decode the B-Format stream to speaker feeds.

Personally I think option 1) would be the best since it is easy to implement, easy to maintain and the most flexible option for possibly a niche feature.

**Comment by ann0see** [comment]
Hi,

Multi channel and probably also surround support was discussed some time ago, I think. The main problem is that the server only supports two channels and changing it would mean a big design change.

**Reply by Hiradur** [reply]
A first step could be that surround sound would only be supported client-side for panning for output only, i.e. no support for surround sound as an input. I'm thinking of the client being able to pan mono streams in a 2D or even 3D sound field (compare with options 2) and 3) in the initial post) or output individual channels for panning in another application (option 1)).

**Reply by pljones** [reply]
The point is the client receives a single stereo pair from the server, regardless of how many other clients are connected to the server, because it's the server doing the mix.  If an "enhanced stereo" mix can be implemented (where "new" clients would remain compatible with "old" servers) that's a possibility, if someone would like to supply the algorithm - and, preferably, a realtime-efficient implementation for up to 150 stereo pairs.

**Reply by Hiradur** [reply]
I see. Alternatively, it would also be possible to transmit surround sound through 2 channels using matrix encoding, e.g. using [Ambisonics UHJ](https://en.wikipedia.org/wiki/Ambisonic_UHJ_format) encoding. This would then require decoding in the client.
[OpenAL Soft, for example, ships with a free implementation of UHJ encoding and decoding](https://github.com/kcat/openal-soft/tree/master/utils).

**Reply by pljones** [reply]
If anyone looks to take this forward:
1. Stick to the current licence or remain compatible with it as a first point (I've not looked at the GitHub code above)
1. Ensure it's the client _requesting_ the up-hanced mix from the server, not a default - but also require that the server have the feature switched on from the command line, to allow the operator to retain control
1. (Again, without looking at the code), remember Jamulus's use case is aiming towards low latency (as a higher priority than a high quality mix) - low buffer sizes for other clients should not be compromised by an up-hanced mix for one (or more ) client

**Reply by gilgongo** [reply]
> 1. Stick to the current licence or remain compatible 

That project is almost as old as Jamulus. I see it's also using the ordinary GNU GPL (rather than the LGPL), which makes it available only for free programs.
